myia/dtype.py

```diff
@@ -0,0 +1,188 @@
+"""Type representation."""
+from typing import Tuple as TupleT, Dict as DictT, Iterable, Any
+from operator import itemgetter
+import collections
+
+KeysT = Iterable[TupleT[str, 'Type']]
+
+
+def make_new(types):
+    """Create a __new__ method with the specified arguments."""
+    args = ', '.join(f'{name}' for name in types.keys())
+
+    ns: DictT[str, Any] = dict()
+    tmpl = f"""def __new__(cls, {args}):
+    return tuple.__new__(cls, ({args}{',' if len(args) != 0 else ''}))
+    """
+    exec(tmpl, ns)
+    new = ns['__new__']
+    new.__annotations__ = types
+    return new
+
+
+def get_types(bases):
+    """Return the type of the class fields."""
+    if len(bases) > 1:
+        raise TypeError("Multiple inheritance not supported")
+    elif len(bases) == 0:
+        return {}
+    elif isinstance(bases[0], TypeMeta):
+        # We need to make a copy here
+        return dict(bases[0]._fields)  # type: ignore
+    return {}
+
+
+class TypeMeta(type):
+    """Metaclass for types.
+
+    Ensures that instances are unique.
+    """
+
+    def __new__(cls, typename, bases, ns):
+        """Create a new Type."""
+        if ns.get('_root', False):
+            bases = (tuple,) + bases
+            ns['_instances'] = dict()
+        if '__slots__' in ns:
+            raise TypeError("Don't define __slots__")
+        ns['__slots__'] = ()
+
+        types = get_types(bases)
+        types.update(filter(lambda v: not v[0].startswith('_'),
+                            ns.get('__annotations__', {}).items()))
+        ns['_fields'] = types
+
+        if '__new__' not in ns:
+            ns['__new__'] = make_new(types)
+        for i, k in enumerate(types.keys()):
+            ns[k] = property(itemgetter(i))
+        return type.__new__(cls, typename, bases, ns)
+
+    def __call__(self, *args, **kwargs):
+        """Overloaded to change the semantics of class instantiation."""
+        if hasattr(self, '_parse_args'):
+            args = self._parse_args(args, kwargs)
+        else:
+            assert len(kwargs) == 0
+        key = (self, args)
+        if key not in self._instances:
+            obj = super().__call__(*args)
+            self._instances[key] = obj
+        return self._instances[key]
+
+
+class Type(metaclass=TypeMeta):
+    """Base class for all Types."""
```

Comment:
Could you document a bit how Type works? Just some high level information that's enough to help someone implement a new Type. From what I understand:

* Type instances are cached so that `Type(x, y, z) is Type(x, y, z)`
* Types are tuples.
* `_parse_args(cls, args, kwargs)` is a class method and serves as an init function. It receives constructor arguments (positional and keyword) which it must check. It must return the values of the attributes declared in the class, in the order that they are declared, except those that start with `_`.

Reply:
done